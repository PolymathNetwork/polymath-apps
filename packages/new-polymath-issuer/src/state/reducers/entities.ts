import {
  createEntityReducer,
  EntityState,
} from '~/state/helpers/createEntityReducer';
import {
  TransactionEntity,
  DividendEntity,
  CheckpointEntity,
  Erc20DividendsModuleEntity,
} from '~/types';
import { combineReducers } from 'redux';

export interface EntitiesState {
  transactions: EntityState<TransactionEntity>;
  checkpoints: EntityState<CheckpointEntity>;
  dividends: EntityState<DividendEntity>;
  erc20DividendsModules: EntityState<Erc20DividendsModuleEntity>;
}

export const reducer = combineReducers<EntitiesState>({
  transactions: createEntityReducer<TransactionEntity>('TRANSACTIONS'),
  checkpoints: createEntityReducer<CheckpointEntity>('CHECKPOINTS'),
  dividends: createEntityReducer<DividendEntity>('DIVIDENDS'),
  erc20DividendsModules: createEntityReducer<Erc20DividendsModuleEntity>(
    'ERC20_DIVIDENDS_MODULES'
  ),
});
