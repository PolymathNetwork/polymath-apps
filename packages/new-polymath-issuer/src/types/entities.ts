export enum TransactionStatus {
  Idle = 'IDLE',
  Unapproved = 'UNAPPROVED',
  Approved = 'APPROVED',
  Rejected = 'REJECTED',
  Succeeded = 'SUCCEEDED',
  Failed = 'FAILED',
}
export enum TransactionGroupStatus {
  Idle = 'IDLE',
  Running = 'RUNNING',
  Failed = 'FAILED',
  Succeeded = 'SUCCEEDED',
}

export type Id = string;

export interface Wallet {
  address: string;
}
export interface Identity {
  id: Id;
  name: string;
  email: string;
}
export interface Transaction {
  id: Id;
  label?: string;
  status: TransactionStatus;
  error?: string; // Maybe an object?
}
export interface TransactionGroup {
  id: Id;
  label?: string;
  status: TransactionGroupStatus;
  transactions: Id[];
  error?: string; // Maybe an object?
}
