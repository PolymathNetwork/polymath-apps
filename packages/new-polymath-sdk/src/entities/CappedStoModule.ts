import BigNumber from 'bignumber.js';
import { Polymath } from '../Polymath';
import { serialize } from '../utils';
import { StoModule, UniqueIdentifiers, Params as StoParams } from './StoModule';

interface Params extends StoParams {
  cap: BigNumber;
  rate: BigNumber;
}

export class CappedStoModule extends StoModule {
  public static generateId({ securityTokenId, stoType, address }: UniqueIdentifiers) {
    return serialize('cappedStoModule', {
      securityTokenId,
      stoType,
      address,
    });
  }

  public uid: string;

  public cap: BigNumber;

  public rate: BigNumber;

  constructor(params: Params, polyClient?: Polymath) {
    const { cap, rate, ...rest } = params;

    super(rest, polyClient);

    const { securityTokenId, address, stoType } = rest;

    this.cap = cap;
    this.rate = rate;
    this.uid = CappedStoModule.generateId({ address, stoType, securityTokenId });
  }

  public toPojo() {
    const stoPojo = super.toPojo();
    const { cap, rate } = this;

    return {
      ...stoPojo,
      cap,
      rate,
    };
  }
}
