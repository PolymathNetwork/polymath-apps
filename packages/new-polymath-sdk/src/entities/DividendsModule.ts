import { Polymath } from '~/Polymath';
import { Entity } from './Entity';

export interface Params {
  address: string;
  securityTokenSymbol: string;
  securityTokenId: string;
  storageWalletAddress: string;
}

export abstract class DividendsModule extends Entity {
  public abstract uid: string;
  public address: string;
  public securityTokenSymbol: string;
  public securityTokenId: string;
  public storageWalletAddress: string;

  constructor(params: Params, polyClient?: Polymath) {
    super(polyClient);

    const {
      address,
      securityTokenSymbol,
      securityTokenId,
      storageWalletAddress,
    } = params;

    this.address = address;
    this.securityTokenSymbol = securityTokenSymbol;
    this.securityTokenId = securityTokenId;
    this.storageWalletAddress = storageWalletAddress;
  }

  public toPojo() {
    const {
      uid,
      address,
      securityTokenSymbol,
      securityTokenId,
      storageWalletAddress,
    } = this;

    return {
      uid,
      address,
      securityTokenSymbol,
      securityTokenId,
      storageWalletAddress,
    };
  }
}
