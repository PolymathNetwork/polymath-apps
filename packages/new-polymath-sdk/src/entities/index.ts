export { SecurityTokenReservation } from './SecurityTokenReservation';
export { SecurityToken } from './SecurityToken';
export { Erc20DividendsModule } from './Erc20DividendsModule';
export { EthDividendsModule } from './EthDividendsModule';
export { Dividend } from './Dividend';
export { Checkpoint } from './Checkpoint';
export { TaxWithholding } from './TaxWithholding';
export { PolyTransaction } from './PolyTransaction';
export { TransactionQueue } from './TransactionQueue';
export { Erc20TokenBalance } from './Erc20TokenBalance';
export { CappedStoModule } from './CappedStoModule';
export { UsdTieredStoModule } from './UsdTieredStoModule';
export { Investment } from './Investment';
export { StoModule } from './StoModule';
