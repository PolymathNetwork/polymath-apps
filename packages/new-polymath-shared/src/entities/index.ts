export * from './Polymath';
export * from './Transaction';
