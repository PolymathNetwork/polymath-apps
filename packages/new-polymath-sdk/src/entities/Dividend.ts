import { Polymath } from '~/Polymath';
import { Entity } from './Entity';
import { serialize } from '~/utils';
import BigNumber from 'bignumber.js';
import { DividendModuleTypes, DividendInvestorStatus } from '~/types';

interface Params {
  index: number;
  checkpointId: string;
  dividendType: DividendModuleTypes;
  securityTokenSymbol: string;
  securityTokenId: string;
  created: Date;
  maturity: Date;
  expiry: Date;
  amount: BigNumber;
  claimedAmount: BigNumber;
  totalSupply: BigNumber;
  reclaimed: boolean;
  totalWithheld: BigNumber;
  totalWithheldWithdrawn: BigNumber;
  investors: DividendInvestorStatus[];
  name: string;
  currency: string | null;
}

export class Dividend extends Entity {
  public static generateId({
    securityTokenSymbol,
    dividendType,
    index,
  }: {
    securityTokenSymbol: string;
    dividendType: DividendModuleTypes;
    index: number;
  }) {
    return serialize('dividend', {
      securityTokenSymbol,
      dividendType,
      index,
    });
  }
  public uid: string;
  public index: number;
  public checkpointId: string;
  public dividendType: DividendModuleTypes;
  public securityTokenSymbol: string;
  public securityTokenId: string;
  public created: Date;
  public maturity: Date;
  public expiry: Date;
  public amount: BigNumber;
  public claimedAmount: BigNumber;
  public totalSupply: BigNumber;
  public reclaimed: boolean;
  public totalWithheld: BigNumber;
  public totalWithheldWithdrawn: BigNumber;
  public investors: DividendInvestorStatus[];
  public name: string;
  public currency: string | null;

  constructor(params: Params, polyClient?: Polymath) {
    super(polyClient);

    const {
      index,
      checkpointId,
      dividendType,
      securityTokenSymbol,
      securityTokenId,
      created,
      maturity,
      expiry,
      amount,
      claimedAmount,
      totalSupply,
      reclaimed,
      totalWithheld,
      totalWithheldWithdrawn,
      investors,
      name,
      currency,
    } = params;

    this.index = index;
    this.checkpointId = checkpointId;
    this.dividendType = dividendType;
    this.securityTokenSymbol = securityTokenSymbol;
    this.securityTokenId = securityTokenId;
    this.created = created;
    this.maturity = maturity;
    this.expiry = expiry;
    this.amount = amount;
    this.claimedAmount = claimedAmount;
    this.totalSupply = totalSupply;
    this.reclaimed = reclaimed;
    this.totalWithheld = totalWithheld;
    this.totalWithheldWithdrawn = totalWithheldWithdrawn;
    this.name = name;
    this.investors = investors;
    this.currency = currency;

    this.uid = Dividend.generateId({
      securityTokenSymbol,
      dividendType,
      index,
    });
  }

  public toPojo() {
    const {
      uid,
      index,
      checkpointId,
      dividendType,
      securityTokenSymbol,
      securityTokenId,
      created,
      maturity,
      expiry,
      amount,
      claimedAmount,
      totalSupply,
      reclaimed,
      totalWithheld,
      totalWithheldWithdrawn,
      investors,
      name,
      currency,
    } = this;

    return {
      uid,
      index,
      checkpointId,
      dividendType,
      securityTokenSymbol,
      securityTokenId,
      created,
      maturity,
      expiry,
      amount,
      claimedAmount,
      totalSupply,
      reclaimed,
      totalWithheld,
      totalWithheldWithdrawn,
      investors,
      name,
      currency,
    };
  }
}
