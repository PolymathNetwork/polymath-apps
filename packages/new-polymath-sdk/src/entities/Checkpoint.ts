import { Polymath } from '~/Polymath';
import { Entity } from './Entity';
import { serialize } from '~/utils';
import { Dividend } from './Dividend';
import { InvestorBalance } from '~/types';
import BigNumber from 'bignumber.js';

interface Params {
  dividends: Dividend[];
  securityTokenSymbol: string;
  securityTokenId: string;
  index: number;
  investorBalances: InvestorBalance[];
  totalSupply: BigNumber;
  createdAt: Date;
}

export class Checkpoint extends Entity {
  public static generateId({
    securityTokenSymbol,
    index,
  }: {
    securityTokenSymbol: string;
    index: number;
  }) {
    return serialize('checkpoint', {
      securityTokenSymbol,
      index,
    });
  }
  public uid: string;
  public dividends: Dividend[];
  public securityTokenSymbol: string;
  public securityTokenId: string;
  public index: number;
  public investorBalances: InvestorBalance[];
  public totalSupply: BigNumber;
  public createdAt: Date;

  constructor(params: Params, polyClient?: Polymath) {
    super(polyClient);

    const {
      dividends,
      securityTokenSymbol,
      securityTokenId,
      index,
      investorBalances,
      totalSupply,
      createdAt,
    } = params;

    this.dividends = dividends;
    this.securityTokenSymbol = securityTokenSymbol;
    this.securityTokenId = securityTokenId;
    this.index = index;
    this.investorBalances = investorBalances;
    this.totalSupply = totalSupply;
    this.createdAt = createdAt;
    this.uid = Checkpoint.generateId({ securityTokenSymbol, index });
  }

  public toPojo() {
    const {
      uid,
      dividends,
      securityTokenSymbol,
      securityTokenId,
      index,
      investorBalances,
      totalSupply,
      createdAt,
    } = this;

    return {
      uid,
      dividends: dividends.map(dividend => dividend.toPojo()),
      securityTokenSymbol,
      securityTokenId,
      index,
      investorBalances,
      totalSupply,
      createdAt,
    };
  }
}
