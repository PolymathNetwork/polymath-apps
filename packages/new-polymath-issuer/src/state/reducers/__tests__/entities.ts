import { reducer } from '../entities';

describe('Reducer: entities', () => {
  test('initialState', () => {
    expect(reducer(undefined, {} as any)).toMatchSnapshot();
  });
});
