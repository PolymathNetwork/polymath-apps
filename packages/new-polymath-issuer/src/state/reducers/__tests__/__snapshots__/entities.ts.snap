// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer: entities initialState 1`] = `
Object {
  "checkpoints": Object {
    "allIds": Array [],
    "byId": Object {},
  },
  "dividends": Object {
    "allIds": Array [],
    "byId": Object {},
  },
  "erc20DividendsModules": Object {
    "allIds": Array [],
    "byId": Object {},
  },
  "erc20TokenBalances": Object {
    "allIds": Array [],
    "byId": Object {},
  },
  "taxWithholdings": Object {
    "allIds": Array [],
    "byId": Object {},
  },
  "transactionQueues": Object {
    "allIds": Array [],
    "byId": Object {},
  },
  "transactions": Object {
    "allIds": Array [],
    "byId": Object {},
  },
}
`;
